platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

```diff
@@ -203,27 +213,259 @@ As a result, `toolChains` should now be configured directly at the top-level of
 
-The default version of TestNG is link:https://github.com/testng-team/testng/releases/tag/7.11.0[7.11.0].
+This behavior can be disabled by setting the `failOnNoDiscoveredTests` property to `false` in the test task configuration:
 
-The default version of Spock is link:https://spockframework.org/spock/docs/2.3/release_notes.html[2.3].
+====
+[.multi-language-sample]
+=====
+.build.gradle.kts
+[source,kotlin]
+----
+tasks.withType<AbstractTestTask>().configureEach {
+    failOnNoDiscoveredTests = false
+}
+----
+=====
+[.multi-language-sample]
+=====
+.build.gradle
+[source,groovy]
+----
+tasks.withType(AbstractTestTask).configureEach {
+    failOnNoDiscoveredTests = false
+}
+----
+=====
+====
+
+==== Stale outputs outside the build directory are no longer deleted
+
+In previous versions of Gradle, class files located outside the build directory were deleted when considered stale. This was a special case for class files registered as outputs of a source set.
+
+Because this setup is uncommon and forced Gradle to eagerly realize all compile related tasks in every build, the behavior has been removed in Gradle 9.
+
+Gradle will continue to clean up stale outputs inside the build directory as needed.
+
+==== `model` and `component` tasks are no longer automatically added
+
+The `model` and `component` tasks report on the structure of legacy software model objects configured for the project.
+Previously, these tasks were automatically added to the project for every build.
+These tasks are now only added to a project when a rule-based plugin is applied (such as those provided by Gradle's support for building native software).
+
+=== Configuration Cache changes
+
+==== Unsupported build event listeners are now configuration cache problems
+
+The build event listener registration method
+link:{javadocPath}/org/gradle/build/event/BuildEventsListenerRegistry.html#onTaskCompletion(org.gradle.api.provider.Provider)[`BuildEventsListenerRegistry.onTaskCompletion`]
+accepts arbitrary providers of any
+link:{javadocPath}/org/gradle/tooling/events/OperationCompletionListener.html[`OperationCompletionListener`]
+implementations.
+However, only providers returned from
+link:{javadocPath}/org/gradle/api/services/BuildServiceRegistry.html#registerIfAbsent(java.lang.String,java.lang.Class,org.gradle.api.Action)[`BuildServiceRegistry.registerIfAbsent`]
+or
+link:{javadocPath}/org/gradle/api/services/BuildServiceRegistration.html#getService()[`BuildServiceRegistration.getService`]
+are currently supported when the Configuration Cache is enabled.
+
+Previously, unsupported providers were silently discarded and never received events when the configuration cache was used.
+Starting with Gradle 9, registering such providers now causes a configuration cache problem and fails the build.
+
+If your build was previously working with the configuration cache (e.g., the listeners were nonessential),
+you can temporarily revert to the old behavior by setting:
```

Comment:
There is a strange-reading mix between listeners and providers in the text. IIUC those are listeners who are incompatible, and the provider just supplies those listeners

Reply:
Will look into this for RC2/GA.

Reply:
> IIUC those are listeners who are incompatible, and the provider just supplies those listeners

You can have incompatible listeners or incompatible providers of compatible listeners. Neither of these actually worked. Compatible providers can only provide compatible listeners.